File: vscode/src/trust/matching.ts
/**
 * Pattern matching for trust patterns.
 * Ported from Python human_review/patterns.py (fnmatch-based matching)
 */

/**
 * Check if a pattern ID matches a glob pattern.
 *
 * Examples:
 *   patternMatchesGlob("imports:added", "imports:*") -> true
 *   patternMatchesGlob("imports:added", "imports:added") -> true
 *   patternMatchesGlob("imports:added", "*:added") -> true
 *   patternMatchesGlob("imports:added", "formatting:*") -> false
 */
export function patternMatchesGlob(patternId: string, globPattern: string): boolean {
  // Exact match
  if (patternId === globPattern) {
    return true;
  }

  // No wildcard, must be exact match
  if (!globPattern.includes("*")) {
    return false;
  }

  // Convert glob pattern to regex
  // Escape special regex chars except *, then convert * to .*
  const escaped = globPattern
    .replace(/[.+?^${}()|[\]\\]/g, "\\$&") // Escape regex special chars
    .replace(/\*/g, ".*"); // Convert * to .*

  const regex = new RegExp(`^${escaped}$`);
  return regex.test(patternId);
}

/**
 * Check if all patterns are trusted by the trust list.
 *
 * @param patterns - List of pattern IDs from a hunk (e.g., ["imports:added"])
 * @param trustList - User's configured trust patterns (e.g., ["imports:*", "formatting:*"])
 * @returns Object with allTrusted boolean and list of untrusted patterns
 */
export function patternsMatchTrustList(
  patterns: string[],
  trustList: string[],
): { allTrusted: boolean; untrustedPatterns: string[] } {
  if (patterns.length === 0) {
    // Empty patterns = needs review (no trustable pattern recognized)
    return { allTrusted: false, untrustedPatterns: [] };
  }

  const untrusted: string[] = [];
  for (const pattern of patterns) {
    const isTrusted = trustList.some((trusted) => patternMatchesGlob(pattern, trusted));
    if (!isTrusted) {
      untrusted.push(pattern);
    }
  }

  return { allTrusted: untrusted.length === 0, untrustedPatterns: untrusted };
}

/**
 * Check if a label matches any pattern in the trust list.
 */
export function isLabelTrusted(label: string, trustList: string[]): boolean {
  return trustList.some((pattern) => patternMatchesGlob(label, pattern));
}

File: src-tauri/src/trust/matching.rs
use glob::Pattern;

/// Check if a label matches a trust pattern (supports wildcards)
///
/// Examples:
/// - "imports:*" matches "imports:added", "imports:removed"
/// - "imports:added" matches only "imports:added"
/// - "*" matches everything
pub fn matches_pattern(label: &str, pattern: &str) -> bool {
    // Convert our pattern syntax to glob syntax
    let glob_pattern = pattern.replace(':', "/");
    let label_path = label.replace(':', "/");

    if let Ok(p) = Pattern::new(&glob_pattern) {
        p.matches(&label_path)
    } else {
        // Fallback to exact match
        label == pattern
    }
}

/// Check if a label matches any pattern in the trust list
pub fn is_trusted(label: &str, trust_list: &[String]) -> bool {
    trust_list
        .iter()
        .any(|pattern| matches_pattern(label, pattern))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exact_match() {
        assert!(matches_pattern("imports:added", "imports:added"));
        assert!(!matches_pattern("imports:added", "imports:removed"));
    }

    #[test]
    fn test_wildcard_match() {
        assert!(matches_pattern("imports:added", "imports:*"));
        assert!(matches_pattern("imports:removed", "imports:*"));
        assert!(!matches_pattern("formatting:whitespace", "imports:*"));
    }

    #[test]
    fn test_is_trusted() {
        let trust_list = vec!["imports:*".to_string(), "formatting:whitespace".to_string()];

        assert!(is_trusted("imports:added", &trust_list));
        assert!(is_trusted("imports:removed", &trust_list));
        assert!(is_trusted("formatting:whitespace", &trust_list));
        assert!(!is_trusted("formatting:style", &trust_list));
    }
}
